packages/base/test/src/widget_test.ts

```diff
@@ -353,6 +353,76 @@ describe('WidgetModel', function () {
       });
       expect(customEventCallback).to.be.calledOnce;
     });
+
+    it('ignores echo_update messages when there is an expected echo_update', async function () {
+      const send = sinon.spy(this.widget, 'send_sync_message');
+      // Set a value, generating an update message, get the message id from the comm?
+      this.widget.set('a', 'original value');
+      this.widget.save_changes();
+
+      // Get the msg id
+      const msgId = send.returnValues[0];
+
+      // Inject a echo_update message from another client
+      await this.widget._handle_comm_msg({
+        parent_header: {
+          msg_id: 'other-client',
+        },
+        content: {
+          data: {
+            method: 'echo_update',
+            state: { a: 'other client update 1' },
+          },
+        },
+      });
+
+      expect(this.widget.get('a')).to.equal('original value');
+
+      // Process a kernel update message, which should set the value
+      await this.widget._handle_comm_msg({
+        parent_header: {
+          msg_id: 'from-kernel',
+        },
+        content: {
+          data: {
+            method: 'update',
+            state: { a: 'kernel update' },
+          },
+        },
+      });
+
+      expect(this.widget.get('a')).to.equal('kernel update');
```

Comment:
The reason we have to have this behavior is that we don't know if we'll ever get our echo back (we don't know if echoing has been disabled for our attribute, or if it has been turned off globally). So I think noting that design decision constraint forces this behavior is probably better than wishing that it didn't happen...

Reply:
Note that I tried to explain this in the protocol description: 

> Since the `echo_update` update messages are optional, and not all attribute updates may be echoed, it is important that only `echo_update` updates are ignored in the last step above, and `update` message updates are always applied.